### DISCUSSION TITLE
Synchronize MIDO with external MIDI-clock

### DISCUSSION THREAD

**Discussion by rolfiepolfie** [mido/mido#379] [Q&A] [answered]
I am planning to write a very simple loop-sequencer using MIDO, and I have an idea how to do this. 
However, I would like to look at the possibility my loop-sequencer can be synchronized from an external clock source  as I got boxes that send TEMPO-messages through their MIDI-out. 

I just want some input how far down the rabbit hole I must go to realize this, like:

1. You can do this relatively easy using MIDO
2. You must use the mido-backend or even code a backend to make this possible
3. you must use another library supporting midi-clock-tempo-sync

Thank you!!

**Comment by rdoursenaud** [comment]
Hi,

Could you please elaborate which part of MIDO you want to sync?
I'd guess MIDI playing but this is out of the scope of MIDO…

Also, you mention `tempo messages`, but these does not exist. There is a `tempo metadata` for MIDI *files* but it's not meant to be sent.
I guess you want to sync your internal tempo to incoming `clock messages` (`OxF8`).
The usual internal tempo representation is `MIDI ticks`.

If that's the case, MIDO may help you code at a higher abstraction  level.
You'll have to code your own synchronization and playback system though.
I'm not aware of any existing library doing this for you.

**Reply by rolfiepolfie** [reply]
As for which part to sync, I am not able to give a you an answer you deserve yet ...
As for my prospect, I guess my "sequencer" will be a simple array/list that gets traversed in a loop, this array will store/record midi values I send to it and play those midi values back when it encounters them in the next pass(es).

As for the  "speed" of this process I now want it to be decided by an external clock-message or external midi-tick (however that works). The only thing I know, that other devices I have like drum-machines are able to be synchronized to midi-"clock", and I want to replicate this way of doing it (my simple external-synced drum-machine does almost the same as I want to be done, when it repeats 1-2 bars endlessly).  

Anyway, I think I need to read myself up  a little ... but that 0xF8 clock message seems the way to go at the moment.

 I just wonder if this might be too much to handle and this project will encounter extreme difficulties if you only have basic python programming skills like me.  Looking status byte of a MIDI message: https://www.recordingblogs.com/wiki/status-byte-of-a-midi-message
 
 I guess this: "0xF8 - Understand the position of the MIDI clock (when synchronized to another device)" are sent quite often to avoid drifting, but I believe just the TEMPO information alone will not help, I need access to the CLOCK-signal itself, (0xFF)?

**Reply by rdoursenaud** [reply]
Thanks for clarifying.

You may want to read the chapter about `System Real Time Messages` of [the official specification](https://www.midi.org/specifications/midi1-specifications/m1-v4-2-1-midi-1-0-detailed-specification-96-1-4) to better understand the concepts involved.

It is indeed not a trivial task.
You have to get an internal timing reference synced to the incoming `MIDI clock` using a PLL (Phase Locked Loop).
And then use that timing reference to play your snippets at the correct tempo.

**Reply by rolfiepolfie** [reply]
Despite being difficult, it's after all the right direction. 
Thank you!!

**Comment by almostimplemented** [comment, marked as answer]
@rdoursenaud Do you have any thoughts on my PR #465 ? I think it is related to this discussion, albeit a much simpler scenario: synchronizing with an arbitrary clock, not dealing with incoming clock messages. (Decided not open a new thread -- sorry if I should have).

My use case is:

I'm timing the firing of MIDI events based on the clock of an external audio interface. I have a `sounddevice.Stream` (wrapper for a PortAudio stream) open on the audio device, so instead of using `time.time()` I use the [stream's `time` property](https://python-sounddevice.readthedocs.io/en/0.3.15/api/streams.html#sounddevice.Stream.time).

**Reply by rolfiepolfie** [reply]
thanks , sorry for late reply!
